### PR TITLE
Adding reserved arguments and pointers in the API to ensure stability.

### PR DESCRIPTION
**Pull request by nicolasnoble**
This is to take care of #1799. The idea is to have extra members in the crucial structures, and extra arguments in the most problematic functions to give us wiggle room later if we need to add more parameters / data.

### CODE REVIEW

File: include/grpc/grpc.h [hunk 1]
```diff
@@ -343,6 +344,7 @@ typedef struct grpc_op {
       int *cancelled;
     } recv_close_on_server;
   } data;
+  void *reserved;
```
Comment (ctiller) [inline include/grpc/grpc.h]: Let's move to before data, and add to the data union:

struct {
  void *reserved[8];
} reserved;

**Comment by ctiller** [comment]
LGTM

**Comment by nicolasnoble** [comment]
Let the compiling-forever-and-fix-stuff game begin...

**Comment by ctiller** [comment]
Unassigned myself until this is revived

**Comment by nicolasnoble** [comment]
This needs reviews / adjustments from the wrapped languages as well.

- [ ] @jtattermusch for C#.
- [ ] @murgatroid99 for node.
- [ ] @tbetbetbe for ruby.
- [ ] @stanley-cheung for PHP.
- [ ] @nathanielmanistaatgoogle / @soltanmm for Python.
- [ ] @jcanizales for Objective-C (sorry, I have little ways to test / fix that, until we get Jenkins/obj-c obviously).

Feel free to submit pull requests on my branch.

**Comment by ctiller** [comment]
Epic.

**Comment by tbetbetbe** [comment]
+1

File: src/node/ext/server.cc [hunk 2]
```diff
@@ -233,7 +233,7 @@ NAN_METHOD(Server::RequestCall) {
       CompletionQueueAsyncWorker::GetQueue(),
       CompletionQueueAsyncWorker::GetQueue(),
       new struct tag(new NanCallback(args[0].As<Function>()), ops.release(),
-                     shared_ptr<Resources>(nullptr)));
+                     shared_ptr<Resources>(NULL)));
```
Comment (murgatroid99) [inline src/node/ext/server.cc]: was this change intentional?

Comment (nicolasnoble) [inline src/node/ext/server.cc]: Nope, good catch. A little search'n'replace went awry.

**Comment by murgatroid99** [comment]
Besides the one issue with `server.cc`, Node changes LGTM

**Comment by nicolasnoble** [comment]
Reverted the server.cc hiccup, and merged master back in to make that PR mergeable again.

**Comment by nicolasnoble** [comment]
Green enough ?

**Comment by vjpai** [comment]
So I generally LGTM the C and C++ code changes that have been made but I have a broader question as to why some functions don't have reserved pointers. For example, grpc_server_request_call and others don't have a new reserved pointer added. Is there a certain class of API functions that are definitely locked down, as opposed to others that might need future expansion?

**Comment by vjpai** [comment]
Oh no, this went conflicted again!

**Comment by soltanmm** [comment]
Did we determine whether or not the channel connectivity functions should get reserved arguments? Or is that for a different PR?

**Comment by nicolasnoble** [comment]
Re: grpc_server_request_call, there's already reserved data in the call object, so there's no need to add one in the arguments. In general, the API changes have been so that anything that doesn't seem to have an already reserved pointer somewhere inside the arguments, we added one. Otherwise, if one of the arguments already has something in them that can facilitate an extra argument, or if the function seems not risky enough, we didn't add any extra argument.

Re: the merge, I'm doing it now. I'm guessing that with the impact of the changes, it'll continue conflicting with almost every other pull request.

Re: channel connectivity, this could be done as a separate PR if needed. This one is already touching enough code like that :-P

**Comment by nicolasnoble** [comment]
Green and mergeable again.

**Comment by nicolasnoble** [comment]
And mergeable again. Let's get that in before it breaks again ?
